rosbag2_to_video/api.py

```diff
@@ -48,120 +62,256 @@ def add_arguments_to_parser(argparser):
+            self._msg_to_cv2 = lambda msg: self._cvbridge.compressed_imgmsg_to_cv2(msg, 'bgr8')
+        storage_filter = rosbag2_py.StorageFilter(topics=[topic_name])
+        self._bag_reader.set_filter(storage_filter)
+    
+    def next(self) -> Tuple['np.ndarray', float]:
+        """Return next image and its timestamp."""
+        _, data, _ = self._bag_reader.read_next()    
+        image_msg = deserialize_message(data, self._msg_type)
+        return self._msg_to_cv2(image_msg), get_stamp_from_image_msg(image_msg)
+    
+    def has_next(self):
+        """Return true if there is at least one more message to read."""
+        return self._bag_reader.has_next()
+
+
+class SequentialVideoWriter:
+    """Create videos from images provided sequentially."""
+
+    def __init__(
+        self,
+        cv_video_writer: cv2.VideoWriter,
+        first_cv_image: 'np.ndarray',
+        start_stamp: float,
+        fps: float,
+    ):
+        """
+        Create a video writer.
+
+        :param cv_video_writer: A cv2.VideoWriter instance, already opened.
+        :param first_cv_image: First image to write in the video.
+        :param start_stamp: Timestamp of the first image.
+        :param fps: Fps used to record the video.
+        """
+        self._fps: float = fps
+        self._cv_video_writer: cv2.VideoWriter = cv_video_writer
+        self._cv_video_writer.write(first_cv_image)
+        self._images_processed: int = 1
+        self._frame_count: int = 1
+        self._images_skipped: int = 0
+        self._start_stamp: float = start_stamp
+        self._last_image: 'np.ndarray' = first_cv_image
+        self._last_image_written_once: bool = True
+    
+    def add_frame(self, cv_image: 'np.ndarray', stamp: float):
+        """Add a frame to the video, given an image and its timestamp."""
+
+        self._images_processed += 1
+        t_from_start = stamp - self._start_stamp
+
+        # accept jitter up to 0.5/fps
+        if t_from_start < (float(self._frame_count) - 0.5) / self._fps:
+            self._last_image = cv_image
+            if not self._last_image_written_once:
+                self._images_skipped += 1
+                print(
+                    'video fps is too low compared to image publish rate, skipping one message',
+                    file=sys.stderr)
+            self._last_image_written_once = False
+            return
+        current_image_frame_index = int(round(t_from_start * self._fps))
+        repeat_last_image = current_image_frame_index - self._frame_count
+        for _ in range(repeat_last_image):
+            self._cv_video_writer.write(self._last_image)
+        self._cv_video_writer.write(cv_image)
+        self._last_image = cv_image
+        if not self._last_image_written_once and repeat_last_image == 0:
+            self._images_skipped += 1
+            print(
+                'video fps is too low compared to image publish rate, skipping one message',
```

Comment:
It would be nice to make this message a little different from the one on L178 so that we can determine which case we're hitting.

Reply:
mmm, for the user both issues are the same.
I'm not sure how to print a different message that's meaningful to the user.

Reply:
Done in https://github.com/ivanpauno/rosbag2_to_video/pull/5.

Reply:
We can leave it as-is, it was just a thought from a debugging perspective.